File: go/3.go
package main 

func main() {
	ss := []string{"abcabcbb", "bbbbb", "pwwkew"}
	for _, s := range ss {
		println(lengthOfLongestSubstring(s))
	}
}

func lengthOfLongestSubstring(s string) int {
	l, r := 0, 0
	set := map[byte]bool{}
	res := 0 

	for r < len(s) {
		for set[s[r]] {
			delete(set, s[l])
			l++
		}
		set[s[r]] = true
		res = max(res, len(set));
		r++
	}
	return res
}

func max(a, b int) int {
	if a > b { return a }
	return b
}

File: cpp/3.cpp
#include <iostream>
#include <tuple>
#include <set>

class Solution
{
public:
	int lengthOfLongestSubstring(std::string s)
	{
		std::set<char> counter;

		int res = 0, l = 0;
		for (int r = 0; r < s.length(); r++)
		{
			while (counter.find(s[r]) != counter.end())
			{
				counter.erase(s[l]);
				l++;
			}
			counter.insert(s[r]);
			if (res < counter.size())
			{
				res = counter.size();
			}
		}

		return res;
	}
};

int main()
{
	std::string ss[] = {"abcabcbb", "bbbbb", "pwwkew"};
	Solution solution;
	for (auto s : ss)
	{
		std::cout << solution.lengthOfLongestSubstring(s) << std::endl;
	}
	return 0;
}